src/Swarm/Game/Robot.hs

```diff
@@ -133,20 +136,40 @@ makeLenses ''LogEntry
 -- | A unique identifier for a robot.
 type RID = Int
 
+-- | The phase of a robot description record.
+data RobotPhase
+  = -- | The robot record has just been read in from a scenario
+    --   description; it represents a /template/ that may later be
+    --   instantiated as one or more concrete robots.
+    TemplateRobot
+  | -- | The robot record represents a concrete robot in the world.
+    ConcreteRobot
+
+-- | With a robot template, we may or may not have a location.  With a
+--   concrete robot we must have a location.
+type family RobotLocation (phase :: RobotPhase) :: * where
+  RobotLocation 'TemplateRobot = Maybe (V2 Int64)
+  RobotLocation 'ConcreteRobot = V2 Int64
+
+-- | Robot templates have no ID; concrete robots definitely do.
+type family RobotID (phase :: RobotPhase) :: * where
+  RobotID 'TemplateRobot = ()
+  RobotID 'ConcreteRobot = RID
+
 -- | A value of type 'RobotR' is a record representing the state of a
 --   single robot.  The @f@ parameter is for tracking whether or not
 --   the robot has been assigned a unique ID.
-data RobotR f = RobotR
+data RobotR (phase :: RobotPhase) = RobotR
```

Comment:
Cool use of `DataKinds`, I like that it's now confined to just two options. :+1:

Reply:
Yeah, I think this is way cleaner than what we had before, even though it uses a bunch more `LANGUAGE` extensions.